### ISSUE TITLE
Correct order of language tag and data type IRI?

### ISSUE THREAD

**Issue by rorlic** [rdfjs/N3.js#307]
It may be a philosophical discussion, but what is the correct order of value, language and datatype in a RDF literal?
A. value, datatype, language (e.g. '"abc"^^http://example.org@en-us')
B. value, language, datatype (e.g. '"abc"@en-us^^http://example.org')

I see option A in the [tests](https://github.com/rdfjs/N3.js/blob/main/test/Term-test.js#L254) but option B. feels more natural and IMHO would be easier to parse because @ could theoretically be part of the IRI.

**Comment by TallTed** [comment]
I believe this is answered in [§3.3 Literals](https://www.w3.org/TR/rdf11-concepts/#section-Graph-Literal) of [_RDF 1.1 Concepts and Abstract Syntax_](https://www.w3.org/TR/rdf11-concepts/), where the order is described as **lexical form, datatype IRI, language tag**.

Also note that the language tag is allowed "if and only if the [datatype IRI](https://www.w3.org/TR/rdf11-concepts/#dfn-datatype-iri) is `http://www.w3.org/1999/02/22-rdf-syntax-ns#langString`" — which is why some serializations don't allow the datatype IRI in a language tagged literal; instead, the presence of a [well-formed language tag](https://www.rfc-editor.org/info/bcp47) on the end of a string literal tells all concerned that the literal is a [`<http://www.w3.org/1999/02/22-rdf-syntax-ns#langString>`](http://www.w3.org/1999/02/22-rdf-syntax-ns#langString).

**Comment by TallTed** [comment]
Also, I suggest you edit your initial posting, and instead of wrapping the sample literals in single-quotes, wrap them in backticks, so that (among other things) you're not tagging the GitHub user `@en-us`, who will otherwise get pinged on every update to this issue.

**Comment by rorlic** [comment]
> Also, I suggest you edit your initial posting, and instead of wrapping the sample literals in single-quotes, wrap them in backticks, so that (among other things) you're not tagging the GitHub user `@en-us`, who will otherwise get pinged on every update to this issue.

Done 😅

**Comment by rorlic** [comment]
> I believe this is answered in [§3.3 Literals](https://www.w3.org/TR/rdf11-concepts/#section-Graph-Literal) of [_RDF 1.1 Concepts and Abstract Syntax_](https://www.w3.org/TR/rdf11-concepts/), where the order is described as **lexical form, datatype IRI, language tag**.
> 
> Also note that the language tag is allowed "if and only if the [datatype IRI](https://www.w3.org/TR/rdf11-concepts/#dfn-datatype-iri) is `http://www.w3.org/1999/02/22-rdf-syntax-ns#langString`" — which is why some serializations don't allow the datatype IRI in a language tagged literal; instead, the presence of a [well-formed language tag](https://www.rfc-editor.org/info/bcp47) on the end of a string literal tells all concerned that the literal is a [`<http://www.w3.org/1999/02/22-rdf-syntax-ns#langString>`](http://www.w3.org/1999/02/22-rdf-syntax-ns#langString).

I checked the specs before but technically **A literal is a language-tagged string if the third element is present.** does not mean the language is in the **last** positon. It is open for interpretation, i.e.
`<value>[@<language>][^^data-type]` would IMHO be a possible and easier option to parse because the `@` is a valid char in an IRI.

All philosophy aside, so it is:
`<value>[^^data-type][@<language>]`

In that case because n3.js does not support it (see above test), the tests should be fixed.

**Comment by TallTed** [comment]
Given that we're discussing `N3.js`, I should perhaps have also included a pointer to [**§ 3.5 Literals**](https://w3c.github.io/N3/spec/#literals) of the latest [_**Notation3** Draft Community Group Report_](https://www.w3.org/standards/types#reports), as of 10 July 2022, which really removes all philosophy (emphasis mine) --

> NOTE
> If no [datatype IRI](https://www.w3.org/TR/xmlschema11-2/#built-in-datatypes) or language tag is given, the datatype `xsd:string` will be assumed. In case a language tag is given, the datatype `rdf:langString` will be assumed. **_Note that it is not possible to specify both a [datatype IRI](https://www.w3.org/TR/xmlschema11-2/#built-in-datatypes) and a language tag._**

This line from the [§ 5.6 Grammar](https://w3c.github.io/N3/spec/#grammar) EBNF should replace your possibilities --

> `[26]   rdfLiteral   ::=   STRING ( LANGTAG | ( "^^" iri) ) ?`

In conclusion -- Yes, the tests should be fixed.
